PSReadLine/Render.cs

```diff
@@ -340,135 +347,330 @@ void MaybeEmphasize(int i, string currColor)
-                        _console.Write("\x1b[0m");
+                        if (logicalLine < renderLines.Length)
+                        {
+                            // This could happen when adding a new line in the end of the very last line.
+                            // In this case, we scroll up by writing out a new line.
+                            _console.SetCursorPosition(left: bufferWidth - 1, top: bufferHeight - 1);
+                            _console.Write("\n");
+                        }
+
+                        // It might happen that 'logicalLine == renderLines.Length'. This means the current
+                        // logical lines to be rendered are exactly the same the the previous logical lines.
+                        // No need to do anything in this case, as we don't need to render anything.
+                    }
+                    else
+                    {
+                        // For the logical line that we will start to re-render from, it's possible that
+                        //   1. the whole logical line had already been scrolled up-off the buffer. This could happen when you backward delete characters
+                        //      on the first line in buffer and cause the current line to be folded to the previous line.
+                        //   2. the logical line spans on multiple physical lines and the top a few physical lines had already been scrolled off the buffer.
+                        //      This could happen when you edit on the top a few physical lines in the buffer, which belong to a longer logical line.
+                        // Either of them will cause 'newTop' to be less than 0.
+                        if (newTop < 0)
+                        {
+                            // In this case, we will render the whole logical line starting from the upper-left-most point of the window.
+                            // By doing this, we are essentially adding a few pseudo physical lines (the physical lines that belong to the logical line but
+                            // had been scrolled off the buffer would be re-rendered). So, update 'physicalLine'.
+                            pseudoPhysicalLineOffset = 0 - newTop;
+                            physicalLine += pseudoPhysicalLineOffset;
+                            newTop = 0;
+                        }
+
+                        _console.SetCursorPosition(left: 0, top: newTop);
                     }
                 }
             }
 
-            int PhysicalLineCount(int columns, bool isFirstLogicalLine, out int lenLastPhysicalLine)
+            if (hasToWriteAll && !cursorMovedToInitialPos)
             {
-                int cnt = 1;
-                if (isFirstLogicalLine)
+                // The editing was in the first logical line. We have to write everything in this case.
+                // Move the cursor to the initial position if we haven't done so.
+                if (_initialY < 0)
```

Comment:
I will go through all the uses of `_initialY`.

Reply:
It will take some time to go through all uses of `_initialY` and verify those uses when `_initialY` is negative. I would like to do this in follow-up PRs. I opened #1015 to track this work.

> because I don't think it could have happened before this PR.

This happens before this PR, but the code never handles this situation properly and that's causing the `ArgumentOutOfRangeException` issues.